### DISCUSSION TITLE
how to build capnp message on the shared momery?

### DISCUSSION THREAD

**Discussion by fengmao31** [capnproto/capnproto#1960] [Q&A]
If I want to use capnp with shared memory, how can I serialized_message on the shared momery.

**Comment by kentonv** [comment]
You would need to write a custom implementation (subclass) of `capnp::MessageBuilder` which allocates segments directly into your shared memory. You may also need a custom implementation of `capnp::MessageReader` to read it.

**Comment by fengmao31** [comment]
1. If I want to request a shared momery block, how to compute the size of the Builder?

2. I read the implainment of MessageReader and messageToFlatArray(), I found it looks like the copy from MessageReader to the 
kj::Array<word> in messageToFlatArray() is not necessary and maybe move is better？
![image](https://github.com/capnproto/capnproto/assets/19343842/2c1479d2-81b6-41db-847d-369e439549c0)

**Reply by kentonv** [reply]
messageToFlatArray() always makes a copy. You probably shouldn't be using that function if you're trying to use shared memory. You should have your custom MessageBuilder allocate segments directly in the shared memory area.

**Reply by fengmao31** [reply]
This is exactly where I'm puzzled; I don't know how to calculate the size of the MessageBuilder class.

**Reply by kentonv** [reply]
You aren't expected to know the size upfront. Messages are allocated in multiple segments -- additional segments are added only as needed. So you'd make a guess for the first segment size, and then if the message grows bigger than that, the system will call back to your implementation to allocate additional segments. Generally I'd use a strategy like allocating each segment to be double the size of the previous.
